clients/drcachesim/reader/snappy_file_reader.h

```diff
@@ -0,0 +1,114 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* snappy_file_reader: reads snappy-compressed files containing memory traces. Files
+ * should follow the snappy framing format:
+ * https://github.com/google/snappy/blob/master/framing_format.txt */
+
+#ifndef _SNAPPY_FILE_READER_H_
+#define _SNAPPY_FILE_READER_H_ 1
+
+#include <fstream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <snappy.h>
+#include <snappy-sinksource.h>
+#include "file_reader.h"
+
+class snappy_reader_t {
+public:
+    snappy_reader_t(std::ifstream *stream);
+
+    // Read 'size' bytes into the 'to'.
+    int
+    read(size_t size, OUT void *to);
+
+    bool
+    eof()
+    {
+        if (!fstream)
+            return true;
+        return fstream->eof();
+    }
+
+private:
+    bool
+    read_new_chunk();
+
+    enum chunk_type_t : unsigned char {
+        COMPRESSED_DATA = 0x00,
+        UNCOMPRESSED_DATA = 0x01,
+        SKIP_BEGIN = 0x80,
+        SKIP_END = 0xfd,
+        PADDING = 0xfe,
+        STREAM_IDENTIFIER = 0xff,
+    };
+
+    // Read a new data chunk in uncompressed_buf.
+    bool
+    read_data_chunk(uint32_t size, chunk_type_t type);
+
+    // Have we seen the magic chunk identifying a snappy stream.
+    bool
+    check_magic();
+
+    // Read and verify magic chunk.
+    bool
+    read_magic(uint32_t size);
+
+    // The compressed file we're reading from.
+    std::unique_ptr<std::ifstream> fstream;
+    // Reader into the decompressed buffer.
+    std::unique_ptr<snappy::Source> src;
+    // Buffer holdling decompressed chunk data.
+    std::vector<char> uncompressed_buf;
+    // BUffer hodling the compressed chunks themselves.
+    std::vector<char> compressed_buf;
+
+    bool seen_magic;
+
+    // Maximum uncompressed chunk size. Fixed by the framing format.
+    static constexpr size_t kMaxBlockSize = 65536;
+    // Maximum compressed chunk size. <= kMaxBlockSize, since the compressor only
+    // emits a compressed chunk if sizes actually shrink.
+    static constexpr size_t kMaxCompressedSize = 65536;
+    // Checksum is always 4 bytes. Buffers should reserve space for it as well.
+    static constexpr size_t kChecksumSize = sizeof(uint32_t);
+    // Magic string identifying the snappy chunked format.
+    static constexpr char kMagic[] = "sNaPpY";
+};
+
+typedef file_reader_t<snappy_reader_t> snappy_file_reader_t;
+
+#endif /* _SNAPPY_FILE_READER_H_ */
```

Comment:
We'll want to install snappy on Travis if it's not already there so Travis will run our snappy test.

Reply:
Done. Checked it wasn't before and it does now.